### PR TITLE
Prevent integer overflow for large sparse matrices on 32-bit systems.

### LINKED ISSUE #37819: Sparse matrix `map` broken for large matrices on 32-bit systems
The following works on a 64-bit system, but fails on a 32-bit system due to integer overflow:

```Julia
using SparseArrays
Z = spzeros(Float64, 50000, 50000)
Z + Z
```
The last line fails with an error like
```
  ArgumentError: invalid Array dimensions
  Stacktrace:
   [1] Array at ./boot.jl:406 [inlined]
   [2] _allocres at /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.5/SparseArrays/src/higherorderfns.jl:233 [inlined]
   [3] _noshapecheck_map(::typeof(+), ::SparseMatrixCSC{Float64,Int32}, ::SparseMatrixCSC{Float64,Int32}) at /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.5/SparseArrays/src/higherorderfns.jl:164
   [4] map at /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.5/SparseArrays/src/higherorderfns.jl:1153 [inlined]
   [5] +(::SparseMatrixCSC{Float64,Int32}, ::SparseMatrixCSC{Float64,Int32}) at /buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.5/SparseArrays/src/sparsematrix.jl:1679
```

This matrix has `0` stored entries and thus fits into memory just fine. However, it has a length of 2,500,000,000 which does not fit into an `Int32`. Integer overflow then leads to a negative number.

The sparse matrix `map` function (`stdlib/SparseArrays/src//higherorderfns.jl`, line 161) contains the line
```Julia
    maxnnzC = fpreszeros ? min(length(A), _sumnnzs(A, Bs...)) : length(A)
```
Since `length(A)` is negative, the new length `maxnnzC` is negative, and this leads to the error shown above.

It seems one solution would be to detect the overflow (e.g. by always using `Int64` in this line), and aborting when `maxnnzC` would be too large for an `Int`.

### PR DESCRIPTION
**Pull request by eschnett**
The length of a sparse array can be very large, even when the number of stored elements is small.
- Add internal "widelength" function for sparse arrays that always uses Int64
- Use it when comparing lenghts
- Add respective tests

Closes https://github.com/JuliaLang/julia/issues/37819.

### CODE REVIEW

**Comment by eschnett** [comment]
The failures are unrelated. The relevant test case `SparseArrays/higherorderfns` succeeded in all runs above. In particular, the relevant 32-bit test suites all succeeded.

File: stdlib/SparseArrays/test/higherorderfns.jl [hunk 1]
```diff
@@ -27,6 +27,9 @@ include("forbidproperties.jl")
         @test map!(cos, X, A) == sparse(map!(cos, fX, fA))
         @test_throws DimensionMismatch map!(sin, X, spzeros((shapeA .- 1)...))
     end
+    # https://github.com/JuliaLang/julia/issues/37819
+    Z = spzeros(50000, 50000)
+    -Z
```
Comment (dkarrasch) [inline stdlib/SparseArrays/test/higherorderfns.jl:32]: Maybe turn this into a little test?
```suggestion
    @test isa(-Z, SparseMatrixCSC{<:Any,Int64})
```
Same below.

Comment (eschnett) [reply stdlib/SparseArrays/test/higherorderfns.jl]: Done.

**Review by dkarrasch — APPROVED** [review]
LGTM.

**Comment by eschnett** [comment]
Ping?

**Pull request merged**